spring-cloud-function-adapters/spring-cloud-function-adapter-gcp/src/test/java/org/springframework/cloud/function/adapter/gcloud/GcfSpringBootHttpRequestHandlerTests.java

```diff
@@ -18,270 +18,125 @@
-		GcfSpringBootHttpRequestHandler<O> handler =
-			new GcfSpringBootHttpRequestHandler<O>(config);
-		this.handler = handler;
-		return handler;
+	@Test
+	public void testHelloWorldSupplier() throws Exception {
+		testFunction(HelloWorldSupplier.class, null, "Hello World!");
 	}
 
 	@Test
-	public void testWithBody() throws Exception {
-		GcfSpringBootHttpRequestHandler<Foo> handler = handler(FunctionMessageBodyConfig.class);
-
-		StringReader foo = new StringReader(GSON.toJson(new Foo("foo")));
-		when(request.getReader()).thenReturn(new BufferedReader(foo));
-
-		StringWriter writer = new StringWriter();
-		handler.service(request, new HttpResponseImpl(new BufferedWriter(writer)));
-
-		assertThat(GSON.fromJson(writer.toString(), Bar.class)).isEqualTo(new Bar("FOO"));
+	public void testJsonInputFunction() throws Exception {
+		testFunction(JsonInputFunction.class, new IncomingRequest("hello"),
+			"Thank you for sending the message: hello");
 	}
 
 	@Test
-	public void testWithRequestParameters() throws Exception {
-		GcfSpringBootHttpRequestHandler<Foo> handler = handler(FunctionMessageEchoReqParametersConfig.class);
-
-		when(request.getReader()).thenReturn(new BufferedReader(new StringReader("")));
-		when(request.getUri()).thenReturn("http://localhost:8080/pathValue");
-		when(request.getPath()).thenReturn("/pathValue");
-		when(request.getHeaders())
-			.thenReturn(Collections.singletonMap("test-header", Collections.singletonList("headerValue")));
-		when(request.getMethod()).thenReturn("GET");
-
-		StringWriter writer = new StringWriter();
-		HttpResponseImpl response = new HttpResponseImpl(new BufferedWriter(writer));
-		handler.service(request, response);
-
-		assertThat(response.statusCode).isEqualTo(200);
-		assertThat(response.headers.get("path")).containsExactly("/pathValue");
-		assertThat(response.headers.get("test-header")).containsExactly("headerValue");
-		assertThat(GSON.fromJson(writer.toString(), Bar.class)).isEqualTo(new Bar("body"));
+	public void testJsonInputOutputFunction() throws Exception {
+		testFunction(JsonInputOutputFunction.class, new IncomingRequest("hello"),
+			new OutgoingResponse("Thank you for sending the message: hello"));
 	}
 
 	@Test
-	public void testWithEmptyBody() throws Exception {
-		GcfSpringBootHttpRequestHandler<Foo> handler = handler(FunctionMessageConsumerConfig.class);
+	public void testJsonInputConsumer() throws Exception {
+		testFunction(JsonInputConsumer.class, new IncomingRequest("hello"), null);
+	}
 
-		when(request.getReader()).thenReturn(new BufferedReader(new StringReader("")));
+	private <I, O> void testFunction(Class<?> configurationClass, I input, O expectedOutput) throws Exception {
+		GcfSpringBootHttpRequestHandler handler = new GcfSpringBootHttpRequestHandler(configurationClass);
 
+		HttpRequest request = Mockito.mock(HttpRequest.class);
+
+		if (input != null) {
+			when(request.getReader()).thenReturn(new BufferedReader(new StringReader(gson.toJson(input))));
```

Comment:
are we sure getReader returns null if we don't send the request body?

Reply:
What do you mean?

Reply:
I mean you don't  set up `getReader` to return anything if  `input != null`, so it would return `null` instead of an empty reader. Are we sure that's what happens if there is no request body?